### PR TITLE
[WIP] Feat : Support Call Kclvm Cli by C API

### PR DESCRIPTION
**Pull request by NeverRaR**
#### 1. Does this PR affect any open issues?(Y/N) and add issue references (e.g. "fix #123", "re #123".):

- [ ] N
- [x] Y 
re #125 

#### 2. What is the scope of this PR (e.g. component or file name):

        KCLVM/kclvm/spec
        KCLVM/kclvm/src/lib.rs

#### 3. Provide a description of the PR(e.g. more details, effects, motivations or doc link):

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [x] Contains experimental features
- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Other

    Support user to call kclvm_cli by c_api

#### 4. Are there any breaking changes?(Y/N) and describe the breaking changes(e.g. more details, motivations or doc link):

- [x] N
- [ ] Y 

#### 5. Are there test cases for these changes?(Y/N) select and add more details, references or doc links:

- [x] Unit test
- [ ] Integration test
- [ ] Benchmark (add benchmark stats below)
- [ ] Manual test (add detailed scripts or steps below)
- [ ] Other

KCLVM/kclvm/src/api_test.rs

#### 6. Release note

Please refer to [Release Notes Language Style Guide](https://kusionstack.io/docs/governance/release-policy/) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by NeverRaR** [comment]
I have read the CLA Document and I hereby sign the CLA

**Review by Peefy — COMMENTED** [review]
Pls add more comments on pub methods/functions/structs and unit/benchmark tests.

File: kclvm/src/lib.rs [hunk 1]
```diff
@@ -4,6 +4,53 @@ use kclvm_parser::load_program;
...
+        unsafe { CString::from_raw(res) };
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn kclvm_service_call(
+    serv: *mut kclvm_service,
+    call: *const c_char,
+    args: *const c_char,
+) -> *const c_char {
```
Comment (chai2010) [inline kclvm/src/lib.rs:41]: add unit test, testdata: https://github.com/KusionStack/KCLVM/tree/main/test/test_units/test_kclvm/test_rpc_server

File: kclvm/spec/src/service/util.rs [hunk 2]
```diff
@@ -0,0 +1,38 @@
+/// New a mutable raw pointer.
+pub fn new_mut_ptr<T>(x: T) -> *mut T {
```
Comment (chai2010) [inline kclvm/spec/src/service/util.rs:2]: dup: https://github.com/KusionStack/KCLVM/blob/main/kclvm/runtime/src/api/utils.rs

File: kclvm/spec/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,26 @@
+[package]
+name = "kclvm-spec"
```
Comment (chai2010) [inline kclvm/spec/Cargo.toml:2]: rename spec to api

File: kclvm/src/kclvm_service_call.h [hunk 4]
```diff
@@ -0,0 +1,25 @@
+#pragma once
+#ifndef KCLVM_SERVICE_CALL
+#define KCLVM_SERVICE_CALL
```
Comment (chai2010) [inline kclvm/src/kclvm_service_call.h:3]: `#define __KCLVM_SERVICE_CALL_H__`

File: kclvm/api/src/service/service.rs [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+                result.yaml_result =
+                    String::from(&yaml_result[yaml_prefix.len()..yaml_result.len()]);
+            } else {
+                result.yaml_result = yaml_result;
+            }
+        }
+        result
+    }
+
+    pub fn reset_plugin(&self, args: &ResetPlugin_Args) -> ResetPlugin_Result {
```
Comment (chai2010) [inline kclvm/api/src/service/service.rs:88]: remove todo methods

**Review by Peefy — COMMENTED** [review]
I suggest to create a new issue and then complete the comments and documentation of `kclvm/api/gpyrpc.proto`, because this is an important user interface. cc @chai2010

**Review by Peefy — COMMENTED** [review]
Benchmark and fatigue tests are required to test for memory leaks and performance issues.

**Review by Peefy — COMMENTED** [review]
U can add --exclude `kclvm/api/src/model/gpyrpc.rs` file in `make codecov` and `make codecov-lcov` at `kclvm/makefile`

File: kclvm/api/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,26 @@
+[package]
+name = "kclvm-api"
+version = "0.1.0"
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
```
Comment (Peefy) [inline kclvm/api/Cargo.toml:6]: remove the comments?

File: kclvm/api/src/service/api.rs [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+pub fn _kclvm_get_service_fn_ptr_by_name(name: &str) -> u64 {
+    match name {
+        "KclvmService.Ping" => ping as *const () as u64,
+        "KclvmService.ExecProgram" => exec_program as *const () as u64,
+        _ => panic!("unknown method name : {}", name),
+    }
+}
+
+pub fn ping(serv: &mut KclvmService, args: &[u8]) -> *const c_char {
```
Comment (Peefy) [inline kclvm/api/src/service/api.rs:16]: more documents on Rust public functions.

File: kclvm/api/src/service/api.rs [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    let args = Ping_Args::parse_from_bytes(args).unwrap();
+    let res = serv.ping(&args);
+    CString::new(res.write_to_bytes().unwrap())
+        .unwrap()
+        .into_raw()
+}
+
+pub fn exec_program(serv: &mut KclvmService, args: &[u8]) -> *const c_char {
```
Comment (Peefy) [inline kclvm/api/src/service/api.rs:24]: more documents on Rust public functions.

File: kclvm/api/src/service/service.rs [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+impl KclvmService {
+    pub fn ping(&self, args: &Ping_Args) -> Ping_Result {
+        Ping_Result {
+            value: (args.value.clone()),
+            special_fields: (args.special_fields.clone()),
+        }
+    }
+
+    pub fn exec_program(&self, args: &ExecProgram_Args) -> ExecProgram_Result {
```
Comment (Peefy) [inline kclvm/api/src/service/service.rs:20]: `exec_program` API and CLI implementation and error handling need to be considered uniformly,

File: kclvm/api/src/service/service.rs [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+            .unwrap();
+            let yaml_prefix = "---\n";
+            if yaml_result.starts_with(yaml_prefix) {
+                result.yaml_result =
+                    String::from(&yaml_result[yaml_prefix.len()..yaml_result.len()]);
+            } else {
+                result.yaml_result = yaml_result;
+            }
+        }
+        result
```
Comment (Peefy) [inline kclvm/api/src/service/service.rs:70]: Put error message in result.

**Review by Peefy — COMMENTED** [review]
Considering the python agent registering cc @chai2010

File: internal/kclvm_py/scripts/build-cpython.sh [hunk 11]
```diff
@@ -35,7 +35,7 @@ do
                 sslpath=$(brew --prefix openssl@1.1)
             fi
 
-            if [ x"$(uname -m)" == x"arm64" ]; then
+            if [ x"$(uname -m)" == x"arm64" ] || [ x"$(uname -m)" == x"x86_64" ]; then
```
Comment (Peefy) [inline internal/kclvm_py/scripts/build-cpython.sh:38]: U can rebase main branch to solve this

**Pull request closed without merging**